benchmark_utils/tv_numba.py

```diff
@@ -0,0 +1,434 @@
+        Return value
+
+    Returns
+    -------
+    np.ndarray
+        Output data.
+
+    References
+    ----------
+    .. [1] https://github.com/albarji/proxTV/blob/master/src/TVL1opt.cpp
+    .. [2] Condat, L. (2013) A direct algorithm for 1D total variation
+    denoising.
+    """
+
+    i = 0
+    mn_height = mx_height = 0
+    mn = y[0] - lmbd
+    mx = y[0] + lmbd
+    last_break = -1
+    mn_break = mx_break = 0
+    N = len(y)
+    while i < N:
+        while i < N - 1:
+            mn_height += mn - y[i]
+            if lmbd < mn_height:
+                i = mn_break + 1
+                x[last_break + 1: mn_break + 1] = mn
+                last_break = mn_break
+                mn = y[i]
+                mx = 2 * lmbd + mn
+                mx_height = lmbd
+                mn_height = -lmbd
+                mn_break = mx_break = i
+                i += 1
+                continue
+            mx_height += mx - y[i]
+            if -lmbd > mx_height:
+                i = mx_break + 1
+                x[last_break + 1: mx_break + 1] = mx
+                last_break = mx_break
+                mx = y[i]
+                mn = mx - 2 * lmbd
+                mn_height = lmbd
+                mx_height = -lmbd
+                mn_break = mx_break = i
+                i += 1
+                continue
+            if mx_height > lmbd:
+                mx += (lmbd - mx_height) / (i - last_break)
+                mx_height = lmbd
+                mx_break = i
+            if mn_height <= -lmbd:
+                mn += (-lmbd - mn_height) / (i - last_break)
+                mn_height = -lmbd
+                mn_break = i
+            i += 1
+        mn_height += mn - y[i]
+        if mn_height > 0:
+            i = mn_break + 1
+            x[last_break + 1: mn_break + 1] = mn
+            last_break = mn_break
+            mn = y[i]
+            mx = 2 * lmbd + mn
+            mx_height = mn_height = -lmbd
+            mn_break = mx_break = i
+            continue
+        mx_height += mx - y[i]
+        if mx_height < 0:
+            i = mx_break + 1
+            x[last_break + 1: mx_break + 1] = mx
+            last_break = mx_break
+            mx = y[i]
+            mn = mx - 2 * lmbd
+            mn_height = mx_height = lmbd
+            mn_break = mx_break = i
+            continue
+        if mn_height <= 0:
+            mn += (-mn_height) / (i - last_break)
+        i += 1
+    x[last_break + 1:] = mn
+    return x
+
+
+linearized_taut_string.jitter = nb.njit(
+    [
+        nb1d32(nbr1d32, nb.types.float32, nb1d32),
+        nb1d64(nbr1d64, nb.types.float64, nb1d64),
+    ]
+)
+
+
+def prox_condat(y, lmbd):
+    x = np.zeros_like(y)
+    linearized_taut_string(y, lmbd, x)
+    return x
+
+
+#######################################
+#   ProxTV 1D using an MM algorithm   #
+#######################################
+
+
+def fast_cost(y, x, r, lmbd):
+    return 0.5 * np.sqrt(np.sum(np.abs(y - x) ** 2)) + lmbd * np.sum(r)
```

Comment:
also `np.linalg.norm(y - x)` is probably a better idea. I am however surprised by the np.sqrt usually the data fit is the squared l2.

Reply:
using the explicit formulation was faster in the micro benchmark I did. 
Regarding the sqrt usage, It was a mistake, thank you. 

I followed the matlab implementations:
 https://eeweb.engineering.nyu.edu/iselesni/lecture_notes/TVDmm/TVD_software/

Reply:
do you have a link to this code in the file? you should keep provenance of the original implementation. To have something even faster then you can do:

```
res = y - x
return 0.5 * np.dot(res, res) + lmbd * np.sum(r)
```

it will avoid some copies.

Reply:
just click tv_mm.m ;) otherwise: 

```octave
function [x, cost] = tvd_mm(y, lam, Nit)
% [x, cost] = tvd_mm(y, lam, Nit)
% Total variation denoising using majorization-minimization
% and banded linear systems.
%
% INPUT
%   y - noisy signal
%   lam - regularization parameter
%   Nit - number of iterations
%
% OUTPUT
%   x - denoised signal
%   cost - cost function history
%
% Reference
% 'On total-variation denoising: A new majorization-minimization
% algorithm and an experimental comparison with wavalet denoising.'
% M. Figueiredo, J. Bioucas-Dias, J. P. Oliveira, and R. D. Nowak.
% Proc. IEEE Int. Conf. Image Processing, 2006.

% Ivan Selesnick, selesi@nyu.edu, 2011
% Revised 2017

y = y(:);                                              % Make column vector
cost = zeros(1, Nit);                                  % Cost function history
N = length(y);

I = speye(N);
D = I(2:N, :) - I(1:N-1, :);
DDT = D * D';

x = y;                                                 % Initialization
Dx = D*x;
Dy = D*y;

for k = 1:Nit
    F = sparse(1:N-1, 1:N-1, abs(Dx)/lam) + DDT;       % F : Sparse banded matrix
    x = y - D'*(F\Dy);                                 % Solve banded linear system
    Dx = D*x;
    cost(k) = 0.5*sum(abs(x-y).^2) + lam*sum(abs(Dx)); % cost function value
end
```
It is referenced in the code.